tsl/src/continuous_aggs/invalidation.c

```diff
@@ -603,19 +680,30 @@ clear_cagg_invalidations_for_refresh(const CaggInvalidationState *state,
 		}
 		else
 		{
-			/* The previous and current invalidation could not be merged. We
-			 * need to cut the prev invalidation against the refresh window */
-			cut_cagg_invalidation(state, refresh_window, &mergedentry);
+			remainder = cut_cagg_invalidation_and_compute_remainder(state,
+																	refresh_window,
+																	&mergedentry,
+																	&remainder);
 			mergedentry = logentry;
 		}
 
 		MemoryContextSwitchTo(oldmctx);
 		MemoryContextReset(state->per_tuple_mctx);
 	}
 
-	/* Handle the last merged invalidation */
+	/* Handle the last (merged) invalidation */
 	if (IS_VALID_INVALIDATION(&mergedentry))
-		cut_cagg_invalidation(state, refresh_window, &mergedentry);
+		remainder = cut_cagg_invalidation_and_compute_remainder(state,
+																refresh_window,
+																&mergedentry,
+																&remainder);
+
+	/* Handle the last (merged) remainder */
+	if (IS_VALID_INVALIDATION(&remainder))
+		save_invalidation(state,
```

Comment:
I suggest to move the check `IS_VALID_INVALIDATION` to inside `save_invalidations`. This is a less error-prone convention.